indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/ParallelIndexSupervisorTask.java

```diff
@@ -1492,31 +1496,57 @@ private RowIngestionMetersTotals getTotalsFromBuildSegmentsRowStats(Object build
-      );
-
-      if (includeUnparseable) {
-        List<ParseExceptionReport> taskUnparsebleEvents = (List<ParseExceptionReport>) reportData.getUnparseableEvents()
-                                                                                   .get(RowIngestionMeters.BUILD_SEGMENTS);
-        unparseableEvents.addAll(taskUnparsebleEvents);
-      }
+      RowIngestionMetersTotals rowIngestionMetersTotals =
+          getBuildSegmentsStatsFromTaskReport(taskReport, includeUnparseable, unparseableEvents);
 
-      processed += totals.getProcessed();
-      processedWithError += totals.getProcessedWithError();
-      thrownAway += totals.getThrownAway();
-      unparseable += totals.getUnparseable();
+      buildSegmentsRowStats.addRowIngestionMetersTotals(rowIngestionMetersTotals);
     }
 
     // Get stats from running tasks
     Set<String> runningTaskIds = parallelSinglePhaseRunner.getRunningTaskIds();
+
+    return getRowStatsAndUnparseableEventsForRunningTasks(runningTaskIds, unparseableEvents, buildSegmentsRowStats, includeUnparseable);
+  }
+
+  private Pair<Map<String, Object>, Map<String, Object>> doGetRowStatsAndUnparseableEventsParallelMultiPhase(boolean includeUnparseable)
+  {
+    if (indexGenerateRowStats != null) {
+      return Pair.of(indexGenerateRowStats.lhs, includeUnparseable ? indexGenerateRowStats.rhs : ImmutableMap.of());
+    } else if (currentSubTaskHolder == null || !((ParallelIndexTaskRunner) currentSubTaskHolder.getTask()).getName().equals("partial segment generation")) {
+      return Pair.of(ImmutableMap.of(), ImmutableMap.of());
+    } else {
+      ParallelIndexTaskRunner currentRunner = currentSubTaskHolder.getTask();
+      Set<String> runningTaskIds = currentRunner.getRunningTaskIds();
+      Map<String, GeneratedPartitionsReport> completedSubtaskReports = (Map<String, GeneratedPartitionsReport>) currentRunner.getReports();
+
+      final MutableRowIngestionMeters buildSegmentsRowStats = new MutableRowIngestionMeters();
+
+      List<ParseExceptionReport> unparseableEvents = new ArrayList<>();
+      for (GeneratedPartitionsReport generatedPartitionsReport : completedSubtaskReports.values()) {
+        Map<String, TaskReport> taskReport = generatedPartitionsReport.getTaskReport();
+        if (taskReport == null || taskReport.isEmpty()) {
+          LOG.warn("Got an empty task report from subtask: " + generatedPartitionsReport.getTaskId());
+          continue;
+        }
+        RowIngestionMetersTotals rowStats =
+            getBuildSegmentsStatsFromTaskReport(taskReport, true, unparseableEvents);
+
+        buildSegmentsRowStats.addRowIngestionMetersTotals(rowStats);
+      }
+      return getRowStatsAndUnparseableEventsForRunningTasks(runningTaskIds, unparseableEvents, buildSegmentsRowStats, includeUnparseable);
+    }
+  }
+
+  private Pair<Map<String, Object>, Map<String, Object>> getRowStatsAndUnparseableEventsForRunningTasks(
```

Comment:
I feel it would be cleaner if this method just returned a `RowIngestionMetersTotals`.

We can have another method which accepts a List of unparseable events and a `MutableRowIngestionMeters` and creates a `Pair<Map<String, Object>, Map<String, Object>>`